kafka-connect-elastic/src/main/scala/com/datamountaineer/streamreactor/connect/elastic/ElasticJsonWriter.scala

```diff
@@ -81,7 +86,25 @@ class ElasticJsonWriter(client: ElasticClient, settings: ElasticSettings) extend
 
         val indexes = sinkRecords
                         .map(r => convert(r, fields, ignoreFields))
-                        .map(r => index into i.toLowerCase / i source r)
+                        .map({ r =>
+                          configMap(r.topic).getWriteMode match {
+                            case WriteModeEnum.INSERT => index into i.toLowerCase / i source r
+                            case WriteModeEnum.UPSERT =>
+                              // Build a Struct field extractor to get the value from the PK field
+                              val pkField = settings.pks(r.topic)
+                              // Extractor includes all since we already converted the records to have only needed fields
+                              val extractor = StructFieldsExtractor(includeAllFields = true, Map(pkField -> pkField))
+                              val fieldsAndValues = extractor.get(r.value.asInstanceOf[Struct]).toMap
+                              if(fieldsAndValues.contains(pkField)){
+                                val pkValue = fieldsAndValues(pkField).toString
+                                update id pkValue in i.toLowerCase / i docAsUpsert fieldsAndValues
+                              } else {
+                                logger.error(s"Empty ${pkField}")
+                                null
```

Comment:
just to clarify - the settings will throw an error if no PK is specified in the properties query, this will throw an error from ES if the PK field is missing from the record - since the id field will be missing on an update.